scripts/install-nix-from-closure.sh

```diff
@@ -1,181 +1,700 @@
 
-echo "performing a single-user installation of Nix..." >&2
+    Doing classic Linux package manager mode.
+    In Nix this mode is called: single-user mode for root.
+
+    Nix can do multi-user mode, and manage package trees for users independently.
+    "
+        fi
+
+        # In case USER is not set
+        # Example: running inside container
+        if [ -z "$USER" ]; then
+            notice "
+
+    Variable USER is not set in the environment.
+    "
+            USER="$(id -u -n)"    # id is POSIX
+            print "Detected username: $USER"
+        fi
+
+        if [ -z "$HOME" ]; then
+            error "
+
+    Environment variable HOME is not set.
+    "
+        fi
+
+    }
+
+    checkHome() {
+        if [ ! -e "$HOME" ]; then
+            error "
+
+    Home directory $HOME does not exist.
+    "
+        fi
+
+        if [ ! -d "$HOME" ]; then
+            error "
+
+    Home directory $HOME is not a directory.
+    "
+        fi
+
+        if [ ! -w "$HOME" ]; then
+            error "
+
+    Home directory $HOME is not writable for user $USER. No deal.
+    "
+        fi
+
+        if [ "$(ls -ld "$HOME" | awk '{print $3}')" != "$USER" ]; then # POSIX, no `test -O`
+            error "
+
+    Home directory $HOME is not owned by user $USER.
+    If you have legitimate case, file a bug with description.
+    So we can handle your particular case.
+    "
+        fi
+
+        if [ ! -x "$HOME" ]; then
+            error "
+
+    Home directory $HOME is not marked as executable for user $USER,
+    how then we are going to go into it?
+    "
+        fi
+
+    }
+
+    checkDest() {
+     if [ -e "$dest" ]; then
+        # Destination directory exist. Nix is/was installed before, be cautious.
+
+        if [ ! -d "$dest" ]; then
+            error "
+
+    '$dest' exists and it's not a directory.
+    "
+        fi
+
+        if [ ! -w $dest ]; then
+            error "
+
+    Directory '$dest' exists, but is not writable for user '$USER'.
+
+    To enable multi-user support see:
+    http://nixos.org/nix/manual/#ssec-multi-user
+
+    To nevertheless do a single-user install for '$USER':
+    recursively set user '$USER' as owner for '$dest' directory.
+    "
+        fi
 
-if ! [ -e $dest ]; then
-    cmd="mkdir -m 0755 $dest && chown $USER $dest"
-    echo "directory $dest does not exist; creating it by running '$cmd' using sudo" >&2
-    if ! sudo sh -c "$cmd"; then
-        echo "$0: please manually run '$cmd' as root to create $dest" >&2
-        exit 1
     fi
-fi
 
-if ! [ -w $dest ]; then
-    echo "$0: directory $dest exists, but is not writable by you. This could indicate that another user has already performed a single-user installation of Nix on this system. If you wish to enable multi-user support see http://nixos.org/nix/manual/#ssec-multi-user. If you wish to continue with a single-user install for $USER please run 'chown -R $USER $dest' as root." >&2
-    exit 1
-fi
+    if [ -e "$dest"/store ]; then
```

Comment:
Introduce a variable for "$dest"/store since you refer to it in many places.

Reply:
`"$dest"/store` is the descriptive name already.
And it used in 4 places.

How can we call it?:

1. `"$store"` - that seems too short to be descriptive. what does that `"$store"` means, where it is.
We also have `"$self/store"`. And we use both `"$self/store"` and `"$dest"/store` during population.

2. Then we should probably use: `"$dest_store"` `"$self_store"` - they already called that, no need to abstract them into variables.

So variable usage is reduced there, and `"$dest"/store` is used.